Make CSV deprecation less annoying to deal with

Girgias opened the pull request:
@nyamsprod I've changed the behaviour of the deprecation to be more sensible in that it requires people to provide the `$escape` parameter.

@cmb69 you might want to have a look too as the original author of https://wiki.php.net/rfc/kill-csv-escaping

cmb69 commented:
I think this makes sense; in the original RFC, I probably should better have written "using" instead of "passing". However, it is almost impossible to get rid of the parameter at some point, since the `$eol` parameter had been added to `fputcsv()`. A similar change to the signature of `imagepolygon()` and friends was only possible due to all parameters have been mandatory anyway.

Girgias commented:
I think it is possible to force usage of a named parameter by checking if the value is set and ZEND_ARG_NUMs is less than its position.

I think the main objective should be to change the default value so that people don't get bitten when writing new code.

cmb69 commented:
> I think the main objective should be to change the default value so that people don't get bitten when writing new code.

Yeah, sure! But it would also be nice to eventually get rid of the implementation of the escape parameter. But that can't happen anytime soon, anyway.

nyamsprod commented:
@Girgias thanks for the update and the change so if I understand correctly the deprecation will only happens if the user explicitly provides an escape parameter other than the empty string

So if no escape parameter is provided or if the default is provided no deprecation should happen correct ?

Girgias commented:
> @Girgias thanks for the update and the change so if I understand correctly the deprecation will only happens if the user explicitly provides an escape parameter other than the empty string
> 
> So if no escape parameter is provided or if the default is provided no deprecation should happen correct ?
> 
> ```
> $obj = new SplTempFileObject();
> $obj->fwrite($csv);
> $obj->setCsvControl(',', escape: "|");  // this should trigger the deprecation notce
> $obj->setCsvControl(',', escape: "\\"); // this should not 
> ```

No, a deprecation is triggered only if you do *not* specify the `$escape` argument, either positionally or with a named argument.

The primary objective is to change the default of the `$escape` parameter from `"\\"` to the empty string `""`.
So this changes would still allow people to use their custom escape argument with PHP's proprietary escape mechanism but they need to explicitly request it.

nyamsprod commented:
OK so if I rewrite my example:

```php
$obj = new SplTempFileObject();
$obj->setCsvControl(',', escape: "|");  // this should not trigger the deprecation notice
$obj->setCsvControl(',', escape: "\\"); // this should not trigger the deprecation notice
$obj->setCsvControl(','); // this SHOULD trigger the deprecation notice
```

Girgias commented:
> OK so if I rewrite my example:
> 
> ```
> $obj = new SplTempFileObject();
> $obj->setCsvControl(',', escape: "|");  // this should not trigger the deprecation notice
> $obj->setCsvControl(',', escape: "\\"); // this should not trigger the deprecation notice
> $obj->setCsvControl(','); // this SHOULD trigger the deprecation notice
> ```

This would be correct yes, I hope this makes it less annoying. :)

jrfnl commented:
> > OK so if I rewrite my example:
> > ```
> > $obj = new SplTempFileObject();
> > $obj->setCsvControl(',', escape: "|");  // this should not trigger the deprecation notice
> > $obj->setCsvControl(',', escape: "\\"); // this should not trigger the deprecation notice
> > $obj->setCsvControl(','); // this SHOULD trigger the deprecation notice
> > ```
> 
> This would be correct yes, I hope this makes it less annoying. :)

I'm not sure that it does and **_it also is no longer in line with the RFC which was voted on_**.

It now requires everyone to start passing the parameter, effectively making the function signature invalid according to PHP itself, as there is now basically a required parameter _after_ two optional parameters.

If we look at prior art, there have been various other functions where the default value of an optional parameter was changed in a PHP minor. To name a few:
* default character set for `htmlspecialchars()`, `htmlentities()` and `html_entity_decode()` in PHP 5.4.
* default character set for `$encoding` for various Iconv and Mbstring functions in PHP 5.6.
* default value for the `$variant` parameter for `idn_to_*()` functions in PHP 7.4.
* default value for the `$flags` parameters for `htmlspecialchars()`, `htmlentities()`, `htmlspecialchars_decode()`, `html_entity_decode()` and `get_html_translation_table()` in PHP 8.1.

None of these changes first made the optional parameter (soft) "required" via a deprecation notice. The default value just changed and people had the choice to pass the parameter if they wanted to enforce either the old or the new behaviour and to not pass the parameter if they didn't care or if the change didn't affect their use-case.

This two-step approach just makes things more complicated.

If I'm honest, I think it would be better to withdraw this change from PHP 8.4 and bring it up for vote again for PHP 8.5 or 9.0, but then to change the default value without a deprecation notice.

nyamsprod commented:
> If I'm honest, I think it would be better to withdraw this change from PHP 8.4 and bring it up for vote again for PHP 8.5 or 9.0, but then to change the default value without a deprecation notice.

I tend to agree with you on that one because the behaviour voted on in the RFC is highly problematic. While the "new" behaviour is much saner I am also in the camp of those who think that changing the parameter default in the next major PHP version without deprecation is the best solution as long as the escape parameter is still around for BC compatibility.

cmb69 commented:
I think, the wording of the RFC was a bit unfortunate (since based on my prior withdrawn RFC). The problem is not really *passing* a non-empty string, but *using* a non-empty string; i.e. the problem is to rely on the proprietary escaping mechanism. As such, the behavior prior to this patch is (almost) useless: those relying on the default will not be notified (although their reliance on the default is likely unconcious), but those deliberately using an `$escape` which is not contained in their data (e .g. `\0`) to circumvent the proprietary escaping will see a deprecation notice.

What about silently changing the default? I think this is somewhat problematic, since CSV may well be used to store persistent data, and if the default changes, that may cause issues when reading CSV written by older PHP versions without any notice about why that happens. Although I have to admit that even writing/reading with the same PHP version may cause issues unless the developer is aware of the proprietary escaping mechanism.

So, in my opinion, we should either stick with what we have now, or completely drop #15362. But this should be discussed on the internals mailing. I'm afraid that many who voted on the RFC are not aware of this PR, but would notice the change when discussed on internals.

jrfnl commented:
> > I'm not sure that it does and **_it also is no longer in line with the RFC which was voted on_**.
> > It now requires everyone to start passing the parameter, effectively making the function signature invalid according to PHP itself, as there is now basically a required parameter _after_ two optional parameters.
> 
> Arguably, the previous implementation was wrong. As the RFC states:
> 
> > [...] is to deprecate _passing_ a non-empty string to the $escape parameter to all the CSV related functions, effectively disabling our problematic escaping mechanism.
> 
> But a default value is still passed to the function, and because the default value is not an empty string it should have, to be correctly implemented, emit a deprecation notice.
> 
> Yes this makes this previously "optional" parameter required, but there is no other way of achieving this, and this argument was used to shut down the RFC from @cmb69 5 years ago, however this still causes issues today and bites people constantly.
> 
> If you want to argue on internals about this, I cannot stop you, but my previous implementation was wrong anyway.

To be honest, I shouldn't have to bring this up on internals. You should though, as you discovered that what was proposed was not a good way forward and you have now implemented a different solution than what was in the RFC.

jrfnl commented:
Thanks @cmb69!

Girgias commented:
> > > I'm not sure that it does and **_it also is no longer in line with the RFC which was voted on_**.
> > > It now requires everyone to start passing the parameter, effectively making the function signature invalid according to PHP itself, as there is now basically a required parameter _after_ two optional parameters.
> > 
> > 
> > Arguably, the previous implementation was wrong. As the RFC states:
> > > [...] is to deprecate _passing_ a non-empty string to the $escape parameter to all the CSV related functions, effectively disabling our problematic escaping mechanism.
> > 
> > 
> > But a default value is still passed to the function, and because the default value is not an empty string it should have, to be correctly implemented, emit a deprecation notice.
> > Yes this makes this previously "optional" parameter required, but there is no other way of achieving this, and this argument was used to shut down the RFC from @cmb69 5 years ago, however this still causes issues today and bites people constantly.
> > If you want to argue on internals about this, I cannot stop you, but my previous implementation was wrong anyway.
> 
> To be honest, I shouldn't have to bring this up on internals. You should though, as you discovered that what was proposed was not a good way forward and you have now implemented a different solution than what was in the RFC.

When I implemented initially, I *thought* it also threw a deprecation on default values of the parameter.
The current implementation of those functions is, questionable.

bukka commented:
So I think this should be reverted as it doesn't match what is in RFC and it's really getting close a real feature freeze.

Also any features (even the RFC implementation) during beta should be approved by (get blessing from) RM (see https://github.com/php/policies/blob/main/release-process.rst#beta-releases ).

bukka commented:
I agree that it makes more sense than the previous implementation and in some way it could be considered as a bug if it didn't change the behaviour so it no longer matches the accepted RFC.

bukka commented:
Maybe we should just revert it all and leave it for the next version...

cmb69 commented:
> if it didn't change the behaviour so it no longer matches the accepted RFC.

As @Girgias [mentioned above](https://github.com/php/php-src/pull/15569#issuecomment-2358768930), this is somewhat of a philosophical question: is a default parameter passed to a function? See also the [only reply to the discussion thread](https://externals.io/message/125628#125629) on internals.

> Maybe we should just revert it all and leave it for the next version...

Like I [stated above](https://github.com/php/php-src/pull/15569#issuecomment-2355265858), either stick with what we have now, or revert it altogether. The original patch is not useful.

nyamsprod commented:
If the new patch is deemed not conform to the original RFC then I would also be in favour of reverting as I also do agree that the original patch was indeed not useful and very difficult to work with from a PHP userland maintainer perspective.

theodorejb commented:
My understanding when reading the RFC was that both not setting an `$escape` value (relying on the default) and explicitly passing a non-empty string as the value would be deprecated.

At the very least, perhaps the current default can be soft-deprecated. I.e. on the manual pages for `fputcsv()`, `fgetcsv()`, and `str_getcsv()`, add a warning that the `$escape` parameter default will change to an empty string in a future PHP release, and this value should be explicitly passed to avoid changes in behavior.

jimwins commented:
Such a warning was actually just pushed earlier today, and should start appearing in the manual shortly.

https://github.com/php/doc-en/pull/3765

Girgias commented:
> So I think this should be reverted as it doesn't match what is in RFC and it's really getting close a real feature freeze.
> 
> Also any features (even the RFC implementation) during beta should be approved by (get blessing from) RM (see https://github.com/php/policies/blob/main/release-process.rst#beta-releases ).

I do not see why I need RM approval here. Moreover, it is not the job of RMs to decide if a change should land in a release, see https://github.com/php/policies/blob/main/release-process.rst#rms-role:

> The roles of the release managers are about being a facilitator:
>
>    [...]
>
> But they are not:
>
>   - Decide which features, extension or SAPI get in a release or not


My original implementation had a bug, as such reverting this would not be sensible, and the original implementation would need to be fixed so that it always throw a deprecation.

I will also say we have routinely amended implementation to be more sensible than what the RFC text said without going through internals or an RM. And I am seriously confused what changed in the 5 years I have actively contributed to PHP to have an obsession with the RFC text rather than the intention of said RFC and the practicalities of the implementation.

bukka commented:
The problem here is that you merged this during beta and we decided that any feature that goes in during beta, needs RM approval. Previously no features would be allowed at all.

If this was in earlier stage, you should still ping internals if someone asks for it and then if there are no objection, you would be free to merge. This is how things always worked AFAIC - we don't require RFC unless there are objections and people request more discussion on internals.

I just think that this change does not leave enough time for any discussion and it's really the last time.

Anyway if RM are fine with that, I'm fine with leaving it but at this stage, this should be decided by RM.

Girgias commented:
A thread has been opened by @cmb69 on internals, where only @kamil-tekiela has replied saying that they thought it would *also* emit a deprecation notice on a default value, thus making my initial implementation wrong.

The only objection has come from @jrfnl about how this effectively makes this parameter "required", which I gather most people would have expected to be the case from the RFC text.

This is the reason why I did not bring this up on internals, because I was working on those assumptions. Now if you want to me to revert the implementation to be more annoying, I can definitely do that, but I don't think it is sensible.

> Previously no features would be allowed at all.

This is incorrect, small features have been added in nearly all beta releases since I have been contributing, moreover the previous text also never mandated that the implementation of RFCs that were voted on needed to be merged prior to Beta1.


Frankly, it feels to me that it has become more and more frustrating and annoying to contribute to PHP in the last 18 months for no apparent reason. And at this point I don't care what happens with this, revert it, leave it, let @php/release-managers-84 decide, let internals have a debate whatever.

We have @nyamsprod, who maintains an incredibly popular library to handle CSV files more easily (because the CSV functions provided by PHP are incredibly bad), who is in favour and is likely one of the main people affected.

But maybe I should just stop wasting my time on fixing those functions and go back to `ext/csv`.

bukka commented:
> This is incorrect, small features have been added in nearly all beta releases since I have been contributing, moreover the previous text also never mandated that the implementation of RFCs that were voted on needed to be merged prior to Beta1.

They have been added but technically they have not been allowed. We were just not respecting that RFC so we added those policies to make it allowed.

So just to clarify, the reason why RM should be involved here is the stability as RM should try to asses if this feature (this is an improvement even though it contains fixes so it is still a feature) is safe and can impact the stability of the final release. Also if there is possibly another problem found, is there a chance this could get fixed / changed in time? 

And what I really meant is to revert the whole thing which means the previously merged one. But as I said, I will leave the decision to RM.

NattyNarwhal commented:
(This is my personal opinion, I'll let Eric and Saki chime in with theirs.)

I think this PR meets the implementation of the RFC; it deprecates passing a non-empty string to $escape as mentioned, and pushes people to get rid of the bad behaviour while keeping existing code working for now. Perhaps the error message and upgrade notice could make it clear what to do to be in good graces (i.e. make it clear setting a non-empty escape is Bad, Actually and escapes will be removed soon, and in upgrade notes, push people to read CSVs with `$escape = default`, write CSVs with `$escape = ""`).

I also don't see the problem with emitting a deprecation message if `$escape` isn't passed. Deprecations aren't errors, after all. That said, I think if the intent is to remove `$escape` in PHP 9, then this and the previous PR encouraging people set $escape in anticipation of getting rid of the current bad default right now means people will have to presumably change their code again to remove the manual override (be it because 8.5 will set a sane default or 9.0 removing it), unless 9.0 will just leave the argument in place but ignore it. So for people who aren't setting `$escape` and don't have any compatibility requirements, this does basically mean a two-step change if they want to not have deprecations in both 8.4 and later 9.0 (or just ignore the 8.4 deprecation message and get the good behaviour change in 9.0). But as previous discussions mention, this might be a requirement to avoid surprises.

I can see the arguments for just changing the default instead of a deprecation message, but I think this PR is compliant with the letter and spirit of the RFC, and sets the stage for improving things. It's a little last-minute to back out the change now, and as Gina suggests, we'd risk bikeshedding how to move forward instead of moving forward.

jimwins commented:
We wont't be able to remove the `escape` parameter everywhere in a future version because it's not the last argument to `fputcsv()` (or `SplFileObject::fputcsv()` any longer. The `eol` parameter  was added in 8.1.0.

cmb69 commented:
> We wont't be able to remove the `escape` parameter everywhere in a future version because it's not the last argument to `fputcsv()` (or `SplFileObject::fputcsv()` any longer. The `eol` parameter was added in 8.1.0.

Right, and that is somewhat unfortunate. However, as soon as `$escape` is fixed (i.e. always an empty string), we can clean up the function, and maybe even fix some bugs – as it is now, it's totally unclear how the functions are even supposed to behave. And we *might* even add proper support for DSV files (like etc/passwd), which I believe was the intention of the `$escape` parameter, but that was messed up with quoting from the beginning. You either quote (CSV) or escape (DSV).
